internal/provider/resource_access_jaas_model_test.go

```diff
@@ -0,0 +1,366 @@
+					resource.TestCheckTypeSetElemAttr(resourceName, "service_accounts.*", svcAccountTwo),
+					resource.TestCheckResourceAttr(resourceName, "users.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "service_accounts.#", "2"),
+				),
+			},
+		},
+	})
+}
+
+// TODO(Kian): Add the test below after a stable release of the provider that includes jaas resources.
+
+// func TestAcc_ResourceJaasAccessModel_UpgradeProvider(t *testing.T) {
+// 	OnlyTestAgainstJAAS(t)
+// 	if testingCloud != LXDCloudTesting {
+// 		t.Skip(t.Name() + " only runs with LXD")
+// 	}
+
+// 	modelName := acctest.RandomWithPrefix("tf-jaas-access-model")
+// 	accessSuccess := "writer"
+
+// 	resourceName := "juju_access_model.test"
+// 	resource.ParallelTest(t, resource.TestCase{
+// 		PreCheck: func() { testAccPreCheck(t) },
+
+// 		Steps: []resource.TestStep{
+// 			{
+// 				ExternalProviders: map[string]resource.ExternalProvider{
+// 					"juju": {
+// 						VersionConstraint: TestProviderStableVersion,
+// 						Source:            "juju/juju",
+// 					},
+// 				},
+// 				Config: testAccResourceJaasAccessModel(modelName, accessSuccess),
+// 				Check: resource.ComposeTestCheckFunc(
+// 					resource.TestMatchResourceAttr(resourceName, "model_uuid", regexp.MustCompile(".+")),
+// 					resource.TestCheckResourceAttr(resourceName, "access", accessSuccess),
+// 					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", "foo@domain.com"),
+// 					resource.TestCheckTypeSetElemAttr(resourceName, "users.*", "bar@domain.com"),
+// 				),
+// 			},
+// 			{
+// 				ProtoV6ProviderFactories: frameworkProviderFactories,
+// 				Config:                   testAccResourceJaasAccessModel(modelName, accessSuccess),
+// 				PlanOnly:                 true,
+// 			},
+// 		},
+// 	})
+// }
+
+func testAccResourceJaasAccessModelTwoUsers(modelName, access, userOne, userTwo string) string {
+	return fmt.Sprintf(`
+resource "juju_model" "test-model" {
+  name = %q
+}
+
+resource "juju_jaas_access_model" "test" {
+  model_uuid   = juju_model.test-model.id
+  access       = %q
+  users        = [%q, %q]
+}`, modelName, access, userOne, userTwo)
+}
+
+func testAccResourceJaasAccessModelOneUser(modelName, access, userOne string) string {
+	return fmt.Sprintf(`
+resource "juju_model" "test-model" {
+  name = %q
+}
+
+resource "juju_jaas_access_model" "test" {
+  model_uuid   = juju_model.test-model.id
+  access       = %q
+  users        = [%q]
+}`, modelName, access, userOne)
+}
+
+func testAccResourceJaasAccessModelOneSvcAccount(modelName, access, svcAccOne string) string {
+	return fmt.Sprintf(`
+resource "juju_model" "test-model" {
+  name = %q
+}
+
+resource "juju_jaas_access_model" "test" {
+  model_uuid          = juju_model.test-model.id
+  access              = %q
+  service_accounts    = [%q]
+}`, modelName, access, svcAccOne)
+}
+
+func testAccResourceJaasAccessModelSvcAccsAndUser(modelName, access, user, svcAccOne, svcAccTwo string) string {
```

Comment:
the preferred plan implementations in tests is to follow the format in testAccResourceSecret, using `internaltesting.GetStringFromTemplateWithData`

https://github.com/juju/terraform-provider-juju/wiki/Developing#writing-acceptance-tests

Reply:
I see, switched to that.